### PR TITLE
Refactor AuditingService base class for external use

### LINKED ISSUE #3594: Refactor AuditingService in NuGetGallery.Core
Refactor the class for referencing by external assembly.

### PR DESCRIPTION
**Pull request by dtivel**
- Fix #3594
- Add a bunch of unit tests to an area that had none, partly to improve test coverage and partly to ensure that the `AuditingService` contracts do not change unexpectedly.
- Rename a few async methods with -`Async` postfix.

### CODE REVIEW

File: tests/NuGetGallery.Core.Facts/Auditing/AuditActorTests.cs [hunk 1]
```diff
@@ -0,0 +1,351 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
```
Comment (skofman1) [inline tests/NuGetGallery.Core.Facts/Auditing/AuditActorTests.cs]: I have one test file for auditing: tests\NuGetGallery.Facts\Infrastructure\UserAuditRecordFacts.cs
Can you please move to NuGetGallery.Core.Facts, so all tests would be in a single project?

**Comment by skofman1** [comment]
:shipit:

File: tests/NuGetGallery.Core.Facts/Auditing/AuditActorTests.cs (hunk 1 above)
Comment (dtivel) [inline tests/NuGetGallery.Core.Facts/Auditing/AuditActorTests.cs]: I think there's value in keeping this test in NuGetGallery.Facts.  This test uses `CredentialBuilder` from NuGetGallery and `TestCredentialBuilder` from NuGetGallery.Facts.  `CredentialBuilder` creates some credentials based on the "latest" version, which is defined by NuGetGallery.  The latest version could change in the future, and this test provides some coverage of that support.

NuGetGallery.Core can't reference NuGetGallery because that would create a circular reference.
